cpp/src/arrow/compute/exec_internal.h

```diff
@@ -72,6 +72,56 @@ class ARROW_EXPORT ExecBatchIterator {
   int64_t max_chunksize_;
 };
 
+/// \brief Break std::vector<Datum> into a sequence of non-owning
+/// ExecSpan for kernel execution. The lifetime of the Datum vector
+/// must be longer than the lifetime of this object
+class ARROW_EXPORT ExecSpanIterator {
+ public:
+  ExecSpanIterator() = default;
+
+  /// \brief Initialize itertor iterator and do basic argument validation
+  ///
+  /// \param[in] batch the input ExecBatch
+  /// \param[in] max_chunksize the maximum length of each ExecSpan. Depending
+  /// on the chunk layout of ChunkedArray.
+  Status Init(const ExecBatch& batch, ValueDescr::Shape output_shape = ValueDescr::ARRAY,
+              int64_t max_chunksize = kDefaultMaxChunksize);
```

Comment:
Nit, but I'm not sure we want to set a default chunksize here? This should be up to the caller, depending on how they want to orchestrate computations.

Reply:
The default here is INT64_MAX, which is to say no chunking